### DISCUSSION TITLE
Custom fallback URL

### DISCUSSION THREAD

**Discussion by ukito-pl** [google/model-viewer#2346] [Ideas]
Hi, 
are there any plans to add option to define custom fallback URL that user will be redirected to if SceneViewer on Android fails to open? 
I can see here: https://github.com/google/model-viewer/blob/6a70645b63f4c8f717b6e5dfb7345ffb071e6462/packages/model-viewer/src/features/ar.ts#L303
that it is currently hard coded to current location url with added hash.
It may become troublesome behavior when using modelviewer via iframe on diffrent domain, beacuse then there is no way to inform user that something went wrong (fallback url is being blocked by browser because domains don't match).

Also it would allow to set fallback url to redirect user to Google Play Store to download or update Google AR Services. It is probably the most common use case for fallback url from what I saw on other solutions.

**Comment by elalish** [comment]
This is an interesting idea; can you put together a Glitch or other repro link where I can repro the problem you're describing?

**Reply by ukito-pl** [reply]
Of course, here is repo when you can see this issue: https://glitch.com/edit/#!/butter-exclusive-lungfish
There is an embedded iframe that loads this page https://smm-bucket.s3.eu-central-1.amazonaws.com/iframe_test/index.html where I show ar-status of model-viewer. If you visit this page directly on your phone that doesn't have Google AR Core installed it will show the status as 'failed'. 
However if you visit this page via iframe like it is done in here https://butter-exclusive-lungfish.glitch.me and do the same the ar-status will not change to 'failed', nothing will happen and user will have no idea what is going on.

Here is another example containing two iframes with only intent links to SceneViewer: https://glitch.com/edit/#!/five-lateral-tile 
First iframe loads webpage with one button that opens SceneViewer in ar mode and fallback url is set to https://developers.google.com/ar. After clicking the button on device without AR Core, user is not being redirected to fallback url. If you acces this webpage directly without iframe, redirection occurs. 
Second iframe is the same but fallback url is set to google play store. In this case user is redirected to google play application regardless if webpage is embedded in iframe or not.

It seems like fallback_url just doesn't work if model-viewer is embedded via iframe with exception of urls that point to google store application. At first I thought that it is because of diffrent domains but in this example https://smm-bucket.s3.eu-central-1.amazonaws.com/iframe_intent_test/index.html host domain and iframe domains are the same and the behavior is still the same. 

In conclusion, it isn't possible to fallback to custom webpage if intent is fired from iframe so my original idea is invalid in this area :( 

But it is still possible to fallback to Google Play Store, so I think that providing `custom_fallback_url` attribute to model-viewer would be beneficial for user experience. Also it would provide additional freedom to developers in designing optimal fallback strategies when using model-viewer in "normal" way without iframes.

**Reply by elalish** [reply]
Thank you for all the detail! I'll have to mull this one over a but. @tpsiaki @AdrianAtGoogle do you have any thoughts on the best way to handle fallback behavior?

**Reply by davidc-epigraph** [reply]
Is there any workaround to implement a fallback URL? curious as having a customer where we have no access to source code (CMS is Magento) but it appears as they something in place that's actively blocking model-viewer from fetching 

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'toString')
    at s.[openSceneViewer] (our-component.min.js:1250:96)
    at s.activateAR (our-component.min.js:1236:18609)
    at XC.[onArButtonClick] (our-component.min.js:1600:5821)
    at Vb.handleEvent (our-component.min.js:1276:7509)
    at HTMLSlotElement.t (field-lounge-chair.html:20:5854)
    at HTMLSlotElement.nrWrapper (field-lounge-chair.html:20:30045)
[openSceneViewer] @ our-component.min.js:1250
activateAR @ our-component.min.js:1236
[onArButtonClick] @ our-component.min.js:1600
handleEvent @ our-component.min.js:1276
t @ product.html:20
nrWrapper @ product.html:20
```

**Reply by elalish** [reply]
Things like this are why we prefer WebXR, which stays in the browser and avoids all this and many other problems. Is there a reason you don't prefer that?

**Reply by tanaydimri** [reply]
Hey @elalish,
Our main reason to not use webXR was the lack of support on iOS with Safari and effectively all other browsers (on iOS). I went to check again on MDN Docs to verify if we missed some updates on this front but looks like iOS is still lacking support for the same. 

I think I should even add that we are only displaying the AR button in this case and  not using the viewer. After looking at the source a bit, it looked like the model viewer component wasn't able to fetch the parent page's location for some reason.

Thank you and we would appreciate any help with this issue.

![image](https://user-images.githubusercontent.com/57341011/164719197-096b3918-7883-481c-8393-fe82f94809a6.png)

**Reply by elalish** [reply]
Fair, but I'm not sure why that matters here. Scene Viewer is equally unsupported on iOS; your only AR option on iOS is Quick Look. On Android your only options are Scene Viewer and WebXR; based on your problem, I'd recommend switching to WebXR. All three of these modes have unique functionality, so there's not really such thing as being more or less consistent. We can only deliver actual consistency when iOS ships WebXR, so please let Apple know you'd like that.

**Reply by tanaydimri** [reply]
Wouldn't we all like it if Apple for once did something in collaboration with the community, sigh!!

Thank you so much for your inputs though, using ar-modes="webxr quick-look" might just do the trick if I understood your correctly. We are trying to have a clone of the client site ready for testing today, but at least setting these ar-modes from the dev console got rid of the error. Nothing launched but that might just be because of the device I was running it on (will try on a mobile phone once our test environment is ready)

Once again, thank you. Really appreciate your quick responses :)
